EL 6 - Adding support for password-auth

nbentoumi opened the pull request:
Created another pull request EL 6 - Adding support for password-auth

Pam config for RHEL6 uses both system-auth and password-auth but the module only manages system-auth.

"The problem with /etc/pam.d/system-auth is that it contains modules that are not usable in remote configurations so remote services such as sshd, vsftpd now use /etc/pam.d/password-auth." https://access.redhat.com/site/solutions/45492.

ghoneycutt commented on README.md:
```diff
@@ -298,6 +298,41 @@ Content template of $system_auth_ac_file. If undef, parameter is set based on th
+
+- *Default*: undef, default is set based on OS version
+
+pam_password_pswd_lines
```
use password instead of pswd

ghoneycutt commented on manifests/init.pp:
```diff
@@ -835,6 +924,8 @@
     create_resources('pam::limits::fragment',$limits_fragments)
   }
 
+  validate_absolute_path($password_auth_ac_file)
```
also need to validate `$password_auth_file` and add the corresponding spec test for failure

ghoneycutt commented:
Would the usage of password-auth also apply to EL7?

ghoneycutt commented on manifests/init.pp:
```diff
@@ -865,26 +956,69 @@
+            }
+            default: {
+
+              file { 'pam_system_auth_ac':
```
instead of duplicating the default in the case statement for EL6, these could live outside of the case, so that they always apply to RedHat regardless of the version.

ghoneycutt commented on manifests/init.pp:
```diff
@@ -827,6 +892,30 @@
   }
 
+  if $pam_password_auth_lines == undef {
+      $my_pam_password_auth_lines = $default_pam_password_auth_lines
```
incorrect spacing. move to the left two spaces.

ghoneycutt commented on manifests/init.pp:
```diff
@@ -827,6 +892,30 @@
+  }
+
+  if $pam_password_account_lines == undef {
+      $my_pam_password_account_lines = $default_pam_password_account_lines
```
incorrect spacing. move to the left two spaces.

ghoneycutt commented on manifests/init.pp:
```diff
@@ -827,6 +892,30 @@
+  }
+
+  if $pam_password_pswd_lines == undef {
+      $my_pam_password_pswd_lines = $default_pam_password_pswd_lines
```
incorrect spacing. move to the left two spaces.

ghoneycutt commented on manifests/init.pp:
```diff
@@ -827,6 +892,30 @@
+  }
+
+  if $pam_password_session_lines == undef {
+      $my_pam_password_session_lines = $default_pam_password_session_lines
```
incorrect spacing. move to the left two spaces.

nbentoumi commented:
For EL7, not sure as i have only the EL6 Redhat KB

nbentoumi commented:
@ghoneycutt 
Travis issue has been fixed and did a commit according to the comments, is there anything need to be done ?

Thanks,
Nasredine

nbentoumi commented:
@ghoneycutt 

I have added support for EL7.

Thanks,
Nasredine

ghoneycutt commented:
This does not seem to make any significant changes to my stock EL7 system, which is as expected. For EL6 I have /etc/pam.d/passwd and /etc/pam.d/password-auth. No password-auth-ac file. I'm not sure the filenames are correct for EL6.

nbentoumi commented:
@ghoneycutt 
The file /etc/pam.d/passwd exists and we need to manage password-auth and password-auth-ac.
RHEL6.5 (Santiago)
ls -l /etc/pam.d/password-auth
lrwxrwxrwx. 1 root root 16 Dec  1 11:55 /etc/pam.d/password-auth -> password-auth-ac

RHEL7.0 (Maipo)
ls -l /etc/pam.d/password-auth
lrwxrwxrwx. 1 root root 16 May 29  2014 /etc/pam.d/password-auth -> password-auth-ac

We have got a request for EL6, and since the pam module supports EL7, i have added EL7 to manage these two files.


Thanks

ghoneycutt commented:
I'm running EL6.6 and not seeing these files. I'm going to look for some documentation on this to make sure everything not just works, but is correct.

Phil-Friderici commented:
@nbentoumi please adjust the indenting of the defaults to match the new style introduced with this commit: https://github.com/ghoneycutt/puppet-module-pam/commit/dd15c26013e0a7e08eba5412d2b91093e85210de

nbentoumi commented:
@ghoneycutt 

Found this documentation about password-auth files in RHEL6
https://access.redhat.com/documentation/en-US/Red_Hat_Enterprise_Linux/6/html/Migration_Planning_Guide/sect-Migration_Guide-Security_Authentication-PAM.html

Phil-Friderici commented:
@nbentoumi thanks for the deep link into the documentation, it confirms your statement:
> Authentication modules are now also written into additional PAM configuration files: /etc/pam.d/password-auth-ac, /etc/pam.d/smartcard-auth-ac and /etc/pam.d/fingerprint-auth-ac.

ghoneycutt commented on README.md:
```diff
@@ -298,6 +298,41 @@ Content template of $system_auth_ac_file. If undef, parameter is set based on th
+
+pam_password_auth_lines
+-------------------------
+Content template of $password_auth_ac_file. If undef, parameter is set based on the OS version.
```
These should say something like the following, so that people know what type the parameter should be.

`Array of lines used in content template of $password_auth_ac_file. If undef, parameter is set based on defaults for the detected platform.`

ghoneycutt commented on manifests/init.pp:
```diff
@@ -827,6 +943,30 @@
     $my_pam_session_lines = $pam_session_lines
   }
 
+  if $pam_password_auth_lines == undef {
```
These all need `validate_array()` after them

ghoneycutt commented on spec/classes/init_spec.rb:
```diff
@@ -598,7 +598,80 @@
+        })
+      }
+      it { should contain_file('pam_password_auth_ac').with_content("# This file is being maintained by Puppet.
+# DO NOT EDIT
```
This should be in a fixture

ghoneycutt commented on spec/classes/init_spec.rb:
```diff
@@ -598,7 +598,80 @@
+    it 'should fail' do
+      expect {
+        should contain_class('pam')
+      }.to raise_error(Puppet::Error)
```
Error's should have a message such as

```ruby
}.to raise_error(Puppet::Error/"invalid\/path" is not an absolute path/)
```

You can find these by putting in a failing spec test and then looking at the output from the tests.
```ruby
}.to raise_error(Puppet::Error/asdf/)
```

ghoneycutt commented on spec/classes/init_spec.rb:
```diff
@@ -831,8 +904,83 @@
+        })
+      }
+      it { should contain_file('pam_password_auth_ac').with_content("# This file is being maintained by Puppet.
+# DO NOT EDIT
```
use a fixture

ghoneycutt commented on spec/classes/limits_spec.rb:
```diff
@@ -75,7 +75,7 @@
           'group'   => 'root',
           'mode'    => '0640',
-          'require' => [ 'Package[pam]', ],
+          'require' => 'Package[pam]',
```
Why does this change?

ghoneycutt commented on spec/classes/init_spec.rb:
```diff
@@ -598,7 +598,80 @@
       }
```
Needs a test for EL5 to show that these files are not present

ghoneycutt replied:
done

ghoneycutt commented on templates/password-auth-ac.erb:
```diff
@@ -0,0 +1,21 @@
+# This file is being maintained by Puppet.
+# DO NOT EDIT
+# Auth
```
Would change slightly for readability

```
# This file is being maintained by Puppet.
# DO NOT EDIT
#
# Auth
```

ghoneycutt replied:
?

Phil-Friderici replied:
@nbentoumi please add this extra line and you have integrated all corrections thath @ghoneycutt mentioned

ghoneycutt replied:
we can skip that as it is purely cosmetic if everything is else ready to merge

ghoneycutt commented:
This would also need to be rebased

treydock commented:
Once this is rebased this change would be a huge improvement.  Right now this module sets sshd to use password-auth, which is currently unmanaged.  My fork uses a crude hack that sets the contents of password-auth-ac to this:

```
# This file is being maintained by Puppet.
# DO NOT EDIT
# Auth
auth       include      system-auth

# Account
account    include      system-auth

# Password
password   include      system-auth

# Session
session    include      system-auth
```

This has worked great until I began testing two-factor authentication which goes in system-auth-ac.  Now commands like "sudo" can't be treated differently than SSH.  So being able to set a different password-auth from system-auth is a huge win.

nbentoumi commented:
@ghoneycutt 
I have updated the pull request, thanks for the comments.
Nasredine

nbentoumi commented:
@ghoneycutt 

I have updated the pull request, thanks for the comments

Nasredine

nbentoumi commented:
@ghoneycutt ,
any comments on this pull request ?

phathocker commented on manifests/init.pp:
```diff
@@ -945,6 +1078,34 @@
+    $my_pam_password_auth_lines = $default_pam_password_auth_lines
+  } else {
+    $my_pam_password_auth_lines = $pam_password_auth_lines
+    validate_array($my_pam_password_auth_lines)
```
@nbentoumi - Wouldn't we just want to validate the $my_pam_password*lines array after each if statement.  That way you validate the default and user input?

ghoneycutt replied:
correct. validate_array() should be outside of the conditional

phathocker commented:
I would love this PR to be merged.  I definitely need the password-auth-ac file to be managed for RHEL6.

ghoneycutt commented on manifests/init.pp:
```diff
@@ -1008,6 +1172,26 @@
             require => Package[$my_package_name],
           }
 
+          if $::operatingsystemmajrelease =~ /^(6|7)$/ {
```
why the regex here?

This seems more straightforward to me.
``` puppet
if $::operatingsystemmajrelease == '6' or $::operatingsystemmajrelease == '7' {
  ...
}
```

ghoneycutt commented on spec/fixtures/pam_password_auth_ac.defaults.el6:
```diff
@@ -0,0 +1,24 @@
+# This file is being maintained by Puppet.
```
The fixtures do not seem to be actually be read and used by the spec tests.

ghoneycutt replied:
?

Phil-Friderici replied:
it is read in line 314 where check can be 'defaults' or 'vas'
`fixture = File.read(fixtures("#{filename}.#{check}.#{k}"))`
`it { should contain_file(filename).with_content(fixture) }``

ghoneycutt commented:
Sorry for the lag, I think we are almost there. Need to address the fixtures being actually used and then have confirmation of this working on actual EL5, 6 and 7 systems. I can confirm on EL6 and 7 and will need help with EL 5 as I do not have ready access to one.

nbentoumi commented:
@ghoneycutt 

I have added SPEC test to read fixture pam_password_auth_ac.{vas,default}.{el6,el7}, and fixed a syntax error on el7 with vas.

Thanks,
Nasredine

ghoneycutt commented:
@nbentoumi you squashed your commits, so I can't tell what you actually changed and have to go through this huge commit from scratch :(

When working with such a big change like this, wait to squash your commits until the merger says it is ready to go, else it creates a lot of re-work.

nbentoumi commented:
@ghoneycutt 

Sorry, was not aware about this process, in my last commit, I have added the code below as suggested 
+            if v[:osfamily] == 'RedHat'
+              if (v[:release] == '6' or v[:release] == '7')
+                  it {
+                    should contain_file('pam_password_auth_ac').with({
+                      'ensure' => 'file',
+                      'path'   => '/etc/pam.d/password-auth-ac',
+                      'owner'  => 'root',
+                      'group'  => 'root',
+                      'mode'   => '0644',
+                    })
+                  }
+                  pam_password_auth_ac_fixture = File.read(fixtures("pam_password_auth_ac.#{check}.#{k}"))
+                  it { should contain_file('pam_password_auth_ac').with_content(pam_password_auth_ac_fixture) }
+              end
+            end


added to $default_pam_password_password_lines the line below on el7 with vas (found this when added the spec test above)
+              'password        requisite       pam_pwquality.so try_first_pass local_users_only retry=3 authtok_type=',

And

added to $default_pam_password_password_lines the line below on el7 with vas 
+              'password        requisite       pam_pwquality.so try_first_pass local_users_only retry=3 authtok_type=',

Phil-Friderici commented:
@ghoneycutt too bad I didn't worked on that, so I don't have an un-squashed branch of this :(
Is there anything else which could help you reviewing this ?

ghoneycutt commented:
Deferring to @Phil-Friderici 

When he says its ready, I'm merging.

Phil-Friderici commented:
enhanced the readability of the comment that @nbentoumi added on Sep 14:

> Sorry, was not aware about this process, in my last commit, I have added the code below as suggested

(added the lines 344 to 358 to init_spec.rb)
>```
            if v[:osfamily] == 'RedHat'
              if (v[:release] == '6' or v[:release] == '7')
                  it {
                    should contain_file('pam_password_auth_ac').with({
                      'ensure' => 'file',
                      'path'   => '/etc/pam.d/password-auth-ac',
                      'owner'  => 'root',
                      'group'  => 'root',
                      'mode'   => '0644',
                    })
                  }
                  pam_password_auth_ac_fixture = File.read(fixtures("pam_password_auth_ac.#{check}.#{k}"))
                  it { should contain_file('pam_password_auth_ac').with_content(pam_password_auth_ac_fixture) }
              end
            end
```
> added to $default_pam_password_password_lines the line below on el7 with vas (found this when
added the spec test above)

(added the line 368 and 428 to init.pp)
>`    'password requisite pam_pwquality.so try_first_pass local_users_only retry=3 authtok_type=',`

anders-larsson commented:
Hello,

I've tested it on the following osfamily/operatingsystemreleas's:
SLE 10.2, 10.3, 10.4, 11.0, 11.1, 11.2, 11.3 11.4 and 12.0
Ubuntu 10.04, 12,04 and 14.04
RHEL 5, 6 and 7
Solaris 10_u8, 10_u9, 10_u10 and 10_u11, 5.11

The changes only seems to affect RHEL systems which is great and it seems to work correctly on systems using VAS/QAS setup. However on systems without VAS/QAS the following change occurs.

/etc/pam.d/password-auth-ac:
{noformat}
-password    sufficient    pam_unix.so md5 shadow nullok try_first_pass use_authtok
+password    sufficient    pam_unix.so sha512 shadow nullok try_first_pass use_authtok
{noformat}

I'm not certain if this is the expected result. Looking at /etc/pam.d/password-auth it seems to use md5 as default. I suspect that a mismatch between this value and the value used in /etc/shadow  (or any other password source) could be devastating?

Please correct me if I'm wrong though.

Phil-Friderici commented:
@anders-larsson Thanks so much for your very helpfull report ! Hope one day you'll introduce and allow me to use your testing rig :)

@nbentoumi I can see that you have added that behaviour to the spec tests as well, so I guess it is done intentionally [1] [2]. Just as @anders-larsson I am wondering why it's different from the usage with VAS included [3] [4].

What's the reason for that ?

[1] [fixture file without vas on RH6](https://github.com/nbentoumi/puppet-module-pam/blob/el6_password_auth_support/spec/fixtures/pam_password_auth_ac.defaults.el6)
[2] [fixture file without vas on RH7](https://github.com/nbentoumi/puppet-module-pam/blob/el6_password_auth_support/spec/fixtures/pam_password_auth_ac.defaults.el7)
[3] [fixture file with vas on RH6](https://github.com/nbentoumi/puppet-module-pam/blob/el6_password_auth_support/spec/fixtures/pam_password_auth_ac.vas.el6)
[4] [fixture file with vas on RH7](https://github.com/nbentoumi/puppet-module-pam/blob/el6_password_auth_support/spec/fixtures/pam_password_auth_ac.vas.el7)

nbentoumi commented:
@anders-larsson ,

Good catch, thanks for the tests, indeed the EL 6 and 7 default use md5, will be working on all comments.
Thanks all

Phil-Friderici commented:
@nbentoumi Looks good to me :+1:  Special thanks for fixing the 'cosmetic change' as well :)

@ghoneycutt this should be finished and mergeable by now

nbentoumi commented:
@ghoneycutt 

is there anything need to be done for this pull request ?

Thanks

Phil-Friderici commented:
@ghoneycutt From my perspective this is mergeable now, what do you think ?

ghoneycutt commented:
@Phil-Friderici OK to merge, but the branch has conflicts and needs to be rebased.

Phil-Friderici commented:
@ghoneycutt thanks for green light
@nbentoumi please rebase to get merged

robert4man commented:
Whatever happened with this? Eagerly awaiting this awesome enhancement. Thanks.

ghoneycutt commented:
strict vars with puppet 4 fails because of nsswitch. Fixing that so this can work.

https://github.com/ghoneycutt/puppet-module-nsswitch/pull/38

nbentoumi commented:
@ghoneycutt 
Thanks, will wait for the branch support_puppet4_strict_vars to be merged.
found also a variable (default_package_name) not defined for Solaris in pam/manifests/init.pp:1065:24, this will not cause spec test to fail for Solaris ?

ghoneycutt commented:
Rebased and watching in PR #129

ghoneycutt commented:
Released in v2.22.0

Big Thanks to everyone who made this happen!!
